# **Predicting Car Insurance Claims Using Machine Learning**

![image](banner_image.png)

## **Project Brief**

Insurance companies invest significant time and resources into accurately assessing customer risk and predicting the likelihood of insurance claims. Effective claim prediction enables insurers to optimise pricing strategies, improve profitability, reduce financial risk, and support better decision-making. Since car insurance is legally required in many countries, the automobile insurance market is highly competitive, making accurate predictive modelling increasingly important.

In this project, *On the Road Car Insurance* requested the development of a simple and interpretable machine learning solution capable of predicting whether a customer would make an insurance claim during the policy period. Due to limited machine learning infrastructure and operational expertise, the company specifically required the identification of a **single feature** that provides the best predictive performance while keeping the final model lightweight and easy to deploy.

The project uses a customer dataset provided in the file `car_insurance.csv`, containing demographic, financial, behavioural, and vehicle-related information such as age, gender, driving experience, education, income level, credit score, vehicle ownership status, vehicle type, driving violations, and accident history. The response variable (`outcome`) indicates whether the customer made an insurance claim.

The primary objective of the project is to:

* Explore and prepare the insurance dataset for modelling.
* Handle missing data appropriately.
* Build individual logistic regression models for each feature.
* Evaluate model performance using classification accuracy.
* Identify the single best predictor of insurance claims.

The project demonstrates the practical application of machine learning and statistical modelling techniques to insurance risk prediction using Python, pandas, NumPy, and the `statsmodels` library.


## **Data Description**

The dataset used in this project is contained in the file `car_insurance.csv`. It includes customer demographic, behavioural, financial, and vehicle-related information collected by the insurance company. The dataset is used to predict whether a customer will make an insurance claim during the policy period.

Each row in the dataset represents a unique customer, while each column represents a specific customer attribute or the target outcome variable.

### Dataset Features

| Column                | Description                                                           |
| --------------------- | --------------------------------------------------------------------- |
| `id`                  | Unique customer identifier                                            |
| `age`                 | Client’s age category                                                 |
| `gender`              | Client’s gender                                                       |
| `driving_experience`  | Number of years the client has been driving                           |
| `education`           | Client’s level of education                                           |
| `income`              | Client’s income category                                              |
| `credit_score`        | Client’s credit score (continuous numerical value between 0 and 1)    |
| `vehicle_ownership`   | Vehicle ownership status                                              |
| `vehicle_year`        | Vehicle registration year category                                    |
| `married`             | Marital status of the client                                          |
| `children`            | Number of children                                                    |
| `postal_code`         | Client’s postal code                                                  |
| `annual_mileage`      | Estimated number of miles driven annually                             |
| `vehicle_type`        | Type of vehicle owned                                                 |
| `speeding_violations` | Total number of speeding violations                                   |
| `duis`                | Number of driving under the influence offences                        |
| `past_accidents`      | Number of previous accidents involving the client                     |
| `outcome`             | Target variable indicating whether the client made an insurance claim |

---

### Feature Categories and Encodings

Several variables in the dataset are encoded numerically to represent categorical information.

#### Age Categories (`age`)

| Value | Category |
| ----- | -------- |
| 0     | 16–25    |
| 1     | 26–39    |
| 2     | 40–64    |
| 3     | 65+      |

#### Gender (`gender`)

| Value | Category |
| ----- | -------- |
| 0     | Female   |
| 1     | Male     |

#### Driving Experience (`driving_experience`)

| Value | Category    |
| ----- | ----------- |
| 0     | 0–9 years   |
| 1     | 10–19 years |
| 2     | 20–29 years |
| 3     | 30+ years   |

#### Education (`education`)

| Value | Category     |
| ----- | ------------ |
| 0     | No education |
| 1     | High school  |
| 2     | University   |

#### Income (`income`)

| Value | Category      |
| ----- | ------------- |
| 0     | Poverty       |
| 1     | Working class |
| 2     | Middle class  |
| 3     | Upper class   |

#### Vehicle Ownership (`vehicle_ownership`)

| Value | Category             |
| ----- | -------------------- |
| 0     | Does not own vehicle |
| 1     | Owns vehicle         |

#### Vehicle Year (`vehicle_year`)

| Value | Category      |
| ----- | ------------- |
| 0     | Before 2015   |
| 1     | 2015 or later |

#### Marital Status (`married`)

| Value | Category    |
| ----- | ----------- |
| 0     | Not married |
| 1     | Married     |

#### Vehicle Type (`vehicle_type`)

| Value | Category   |
| ----- | ---------- |
| 0     | Sedan      |
| 1     | Sports car |

---

### Target Variable

The response variable for this project is `outcome`.

| Value | Description             |
| ----- | ----------------------- |
| 0     | No insurance claim made |
| 1     | Insurance claim made    |

The modelling objective is to determine which single feature best predicts the likelihood of a customer making an insurance claim.


In [1]:
# Import required modules
import pandas as pd
import numpy as np
from statsmodels.formula.api import logit

# Start coding!

## **Task 1: Reading in and exploring the dataset**
Create a pandas DataFrame and examine for data types, missing values, and distributions

### Exploring the data
- Read in the dataset as a pandas DataFrame using pd.read_csv().
- Explore the data to see what the values look like, the data types, missing values, and distributions.
- You can use methods such as .head(), .info(), and .describe().

In [2]:
# Load the CSV file into a pandas DataFrame
car_insurance_df = pd.read_csv("car_insurance.csv")

# Display the first few rows

print("First 5 rows of the dataset:\n")
display(car_insurance_df.head())


First 5 rows of the dataset:



,id,age,gender,driving_experience,education,income,credit_score,vehicle_ownership,vehicle_year,married,children,postal_code,annual_mileage,vehicle_type,speeding_violations,duis,past_accidents,outcome
0,569520,3,0,0-9y,high school,upper class,0.629027,1.0,after 2015,0.0,1.0,10238,12000.0,sedan,0,0,0,0.0
1,750365,0,1,0-9y,none,poverty,0.357757,0.0,before 2015,0.0,0.0,10238,16000.0,sedan,0,0,0,1.0
2,199901,0,0,0-9y,high school,working class,0.493146,1.0,before 2015,0.0,0.0,10238,11000.0,sedan,0,0,0,0.0
3,478866,0,1,0-9y,university,working class,0.206013,1.0,before 2015,0.0,1.0,32765,11000.0,sedan,0,0,0,0.0
4,731664,1,1,10-19y,none,working class,0.388366,1.0,before 2015,0.0,0.0,32765,12000.0,sedan,2,0,1,1.0


In [3]:
# Examine dataset structure and data types

print("\nDataset information:\n")
print(car_insurance_df.info())


Dataset information:

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   10000 non-null  int64  
 1   age                  10000 non-null  int64  
 2   gender               10000 non-null  int64  
 3   driving_experience   10000 non-null  str    
 4   education            10000 non-null  str    
 5   income               10000 non-null  str    
 6   credit_score         9018 non-null   float64
 7   vehicle_ownership    10000 non-null  float64
 8   vehicle_year         10000 non-null  str    
 9   married              10000 non-null  float64
 10  children             10000 non-null  float64
 11  postal_code          10000 non-null  int64  
 12  annual_mileage       9043 non-null   float64
 13  vehicle_type         10000 non-null  str    
 14  speeding_violations  10000 non-null  int64  
 15  duis                 1000

In [4]:
# Check for missing values

print("\nMissing values in each column:\n")
print(car_insurance_df.isnull().sum())


Missing values in each column:

id                       0
age                      0
gender                   0
driving_experience       0
education                0
income                   0
credit_score           982
vehicle_ownership        0
vehicle_year             0
married                  0
children                 0
postal_code              0
annual_mileage         957
vehicle_type             0
speeding_violations      0
duis                     0
past_accidents           0
outcome                  0
dtype: int64


In [5]:
# Summary statistics for numerical columns

print("\nSummary statistics for numerical columns:\n")
display(car_insurance_df.describe())


Summary statistics for numerical columns:



,id,age,gender,credit_score,vehicle_ownership,married,children,postal_code,annual_mileage,speeding_violations,duis,past_accidents,outcome
count,10000.000000,10000.000000,10000.000000,9018.000000,10000.000000,10000.000000,10000.000000,10000.000000,9043.000000,10000.000000,10000.00000,10000.000000,10000.000000
mean,500521.906800,1.489500,0.499000,0.515813,0.697000,0.498200,0.688800,19864.548400,11697.003207,1.482900,0.23920,1.056300,0.313300
std,290030.768758,1.025278,0.500024,0.137688,0.459578,0.500022,0.463008,18915.613855,2818.434528,2.241966,0.55499,1.652454,0.463858
min,101.000000,0.000000,0.000000,0.053358,0.000000,0.000000,0.000000,10238.000000,2000.000000,0.000000,0.00000,0.000000,0.000000
25%,249638.500000,1.000000,0.000000,0.417191,0.000000,0.000000,0.000000,10238.000000,10000.000000,0.000000,0.00000,0.000000,0.000000
50%,501777.000000,1.000000,0.000000,0.525033,1.000000,0.000000,1.000000,10238.000000,12000.000000,0.000000,0.00000,0.000000,0.000000
75%,753974.500000,2.000000,1.000000,0.618312,1.000000,1.000000,1.000000,32765.000000,14000.000000,2.000000,0.00000,2.000000,1.000000
max,999976.000000,3.000000,1.000000,0.960819,1.000000,1.000000,1.000000,92101.000000,22000.000000,22.000000,6.00000,15.000000,1.000000


In [6]:
# Summary statistics for categorical/discrete columns

print("\nSummary statistics for categorical columns:\n")
display(car_insurance_df.describe(include='object'))


Summary statistics for categorical columns:



C:\Users\midem\AppData\Local\Temp\ipykernel_21784\1198842908.py:4: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  display(car_insurance_df.describe(include='object'))


,driving_experience,education,income,vehicle_year,vehicle_type
count,10000,10000,10000,10000,10000
unique,4,3,4,2,2
top,0-9y,high school,upper class,before 2015,sedan
freq,3530,4157,4336,6967,9523


In [ ]:
# Examine value distributions for each column

print("\nValue distributions:\n")

for column in car_insurance_df.columns:
    print(f"\nColumn: {column}")
    display(car_insurance_df[column].value_counts(dropna=False))

In [7]:
# Display dataset shape

print("\nDataset shape:")
print(f"Rows: {car_insurance_df.shape[0]}")
print(f"Columns: {car_insurance_df.shape[1]}")


Dataset shape:
Rows: 10000
Columns: 18


In [8]:
# Check duplicated rows

print("\nNumber of duplicated rows:")
print(car_insurance_df.duplicated().sum())


Number of duplicated rows:
0


## **Task 2: Filling missing values**
Prepare data for modeling by ensuring there are no missing values.

### Filling missing values with an appropriate statistic
- Two of the columns have missing values, and their values appear to be normally distributed.
- Replace missing values in these columns with the mean.
- You can use the Series .fillna() method to fill missing values, passing the .mean() of the respective column.

In [9]:
# Fill missing values with the column mean
car_insurance_df["credit_score"] = car_insurance_df["credit_score"].fillna(
            car_insurance_df["credit_score"].mean()
        )

car_insurance_df["annual_mileage"] = car_insurance_df["annual_mileage"].fillna(
            car_insurance_df["annual_mileage"].mean()
        )

# Verify missing values have been filled
print("\nMissing values after filling:\n")
print(car_insurance_df.isnull().sum())
#display(car_insurance_df)



Missing values after filling:

id                     0
age                    0
gender                 0
driving_experience     0
education              0
income                 0
credit_score           0
vehicle_ownership      0
vehicle_year           0
married                0
children               0
postal_code            0
annual_mileage         0
vehicle_type           0
speeding_violations    0
duis                   0
past_accidents         0
outcome                0
dtype: int64


## **Task 3: Preparing for modelling**
Create variables for modeling and storing the results.

### Creating a list to store the models
- Create an empty list called models to store each model object created.
- You can create an empty list using square brackets, e.g., a_list = [].

### Storing the features as a variable
- Create a variable called features, containing all columns except for "outcome" and "id".
- To get all the features you can chain the .drop() method, passing a list of columns to drop, with the .columns attribute.

In [10]:
# Create an empty list to store model objects
models = []

# Store feature column names
# Select all columns except "outcome" and "id"
features = car_insurance_df.drop(
    columns=["outcome", "id"]
).columns

# Display the features
print("\nFeature columns:\n")
print(features)


Feature columns:

Index(['age', 'gender', 'driving_experience', 'education', 'income',
       'credit_score', 'vehicle_ownership', 'vehicle_year', 'married',
       'children', 'postal_code', 'annual_mileage', 'vehicle_type',
       'speeding_violations', 'duis', 'past_accidents'],
      dtype='str')


## **Task 4: Building and storing the models**
Build one model per feature and save the results to a list.

### Modeling with a for loop
- Loop through features.
- Inside the for loop, create a Logistic Regression model to estimate the relationship between "outcome" and the iterator from features, fitting the data, and saving as a variable called models.
- You can use an f-string to pass a variable inside a string through the use of curly brackets {}. For example, to print the name of the iterator feature in a for loop:

```bash
for col in features:
  print(f"{col}"")
```

- Append the model to the models list.

In [11]:
# Build one logistic regression model per feature
for col in features:
    
    # Display the feature currently being modeled
    print(f"Building model for feature: {col}")
    
    # Create logistic regression formula
    formula = f"outcome ~ {col}"
    
    # Build and fit the logistic regression model
    model = logit(formula=formula, data=car_insurance_df).fit()
    
    # Append the fitted model to the models list
    models.append(model)

# ------------------------------------------------------------
# Confirm models have been stored
# ------------------------------------------------------------

print("\nNumber of models created:")
print(len(models))
#print(models)

Building model for feature: age
Optimization terminated successfully.
         Current function value: 0.511794
         Iterations 6
Building model for feature: gender
Optimization terminated successfully.
         Current function value: 0.615951
         Iterations 5
Building model for feature: driving_experience
Optimization terminated successfully.
         Current function value: 0.467092
         Iterations 8
Building model for feature: education
Optimization terminated successfully.
         Current function value: 0.603742
         Iterations 5
Building model for feature: income
Optimization terminated successfully.
         Current function value: 0.531499
         Iterations 6
Building model for feature: credit_score
Optimization terminated successfully.
         Current function value: 0.572557
         Iterations 6
Building model for feature: vehicle_ownership
Optimization terminated successfully.
         Current function value: 0.552412
         Iterations 5
Building mod

## **Task 5: Measuring performance**
Calculate the accuracy of each model.

### Creating a list to store model accuracies
- Create an empty list and store it as accuracies.

### Looping through the index of the models list
- Iterate over the index of models.
- The iterable should be a list of numbers from 0 to the maximum index in models, which can be accomplished using range(len(models)).

### Creating a confusion matrix and storing individual metrics
- Store each model's confusion matrix by selecting the ith model from models and call the pred_table() method, saving as conf_matrix.
- Find and store the true negatives, true positives, false negatives, and false positives from conf_matrix.
- The true negatives are the top-left value, true positives are the bottom right, false negatives are the bottom left, and false positives are the top right.
- To extract the true negatives use the syntax conf_matrix[0,0].


### Calculating accuracy and storing the model's results
- Calculate accuracy, which is the sum of true negatives and positives divided by the sum of all values in the confusion matrix.
- Append the model's accuracy to the accuracies list.

In [12]:
# Create an empty list to store model accuracies
accuracies = []

# ------------------------------------------------------------
# Loop through all models and calculate accuracy
# ------------------------------------------------------------

for i in range(len(models)):
    
    # Generate confusion matrix for the model
    conf_matrix = models[i].pred_table()
    
    # Extract confusion matrix values
    true_negatives = conf_matrix[0, 0]
    false_positives = conf_matrix[0, 1]
    false_negatives = conf_matrix[1, 0]
    true_positives = conf_matrix[1, 1]
    
    # Calculate model accuracy
    accuracy = (
        true_negatives + true_positives
    ) / conf_matrix.sum()
    
    # Append accuracy to the accuracies list
    accuracies.append(accuracy)

# ------------------------------------------------------------
# Display model accuracies
# ------------------------------------------------------------

print("\nModel Accuracies:\n")

for feature, accuracy in zip(features, accuracies):
    print(f"{feature}: {accuracy:.4f}")


Model Accuracies:

age: 0.7747
gender: 0.6867
driving_experience: 0.7771
education: 0.6867
income: 0.7425
credit_score: 0.7054
vehicle_ownership: 0.7351
vehicle_year: 0.6867
married: 0.6867
children: 0.6867
postal_code: 0.6867
annual_mileage: 0.6904
vehicle_type: 0.6867
speeding_violations: 0.6867
duis: 0.6867
past_accidents: 0.6867


## **Task 6: Finding the best performing model**
Locate which model has the highest accuracy score.

### Identifying the index of accuracies with the largest score
- You can find the index of an item in a list by calling the list's .index() method.
- Inside .index(), you need to look up the max() value in accuracies.

### Mapping the highest accuracy to the feature
- Slice features using square brackets [] containing the index of the model in accuracies with the highest score.
- Use .pd.DataFrame() to create a pandas DataFrame, passing a dictionary with "best_feature" and "best_accuracy" as keys, along with the feature and its accuracy score as values.
- Set the index keyword argument equal to a list containing 0.

In [13]:
# Find the highest accuracy score
max_accuracy = max(accuracies)

# Find the index of the best performing model
best_model_index = accuracies.index(max_accuracy)

# Identify the best feature
best_feature = features[best_model_index]

# Create the final DataFrame
best_feature_df = pd.DataFrame(
    {
        "best_feature": [best_feature],
        "best_accuracy": [max_accuracy]
    },
    index=[0]
)

# Display the result
print("\nBest Performing Feature:\n")
print(best_feature_df)


Best Performing Feature:

         best_feature  best_accuracy
0  driving_experience         0.7771


## **Conclusion**

This project successfully developed a simple machine learning approach for predicting whether a customer is likely to make a car insurance claim. Using logistic regression, individual models were built for each feature in the dataset to determine which single variable provided the highest predictive accuracy for the target variable (`outcome`).

During the project, the dataset was explored and assessed for data quality, including data types, missing values, distributions, and duplicate records. Missing values in the numerical columns were handled using mean imputation to ensure the dataset was suitable for modelling.

A separate logistic regression model was then created for each feature, and the predictive performance of each model was evaluated using classification accuracy derived from the confusion matrix. The results showed that the feature **`driving_experience`** produced the highest accuracy score of approximately **77.71%**, making it the strongest single predictor of whether a customer would submit an insurance claim.

The findings suggest that a customer’s driving experience has a significant relationship with insurance claim likelihood and can serve as an effective low-complexity predictor for insurance risk assessment. This aligns with real-world insurance practices, where driving history and experience are commonly used as key indicators of risk.

Overall, the project demonstrates how relatively simple machine learning techniques and careful feature evaluation can provide meaningful business insights while maintaining low implementation complexity. Such lightweight models are especially useful for organisations with limited machine learning infrastructure and deployment capabilities.
